src/mame/midway/midyunit.h

```diff
@@ -115,44 +87,140 @@ class midyunit_state : public driver_device
-	void midyunit_control_w(offs_t offset, uint16_t data, uint16_t mem_mask = ~0);
-	void midyunit_paletteram_w(offs_t offset, uint16_t data, uint16_t mem_mask = ~0);
-	uint16_t midyunit_dma_r(offs_t offset);
-	void midyunit_dma_w(offs_t offset, uint16_t data, uint16_t mem_mask = ~0);
+	void cmos_w(offs_t offset, uint16_t data, uint16_t mem_mask = ~0);
+	uint16_t cmos_r(offs_t offset);
+	void cmos_enable_w(address_space &space, uint16_t data);
+	uint16_t protection_r();
+	uint16_t input_r(offs_t offset);
+	uint16_t gfxrom_r(offs_t offset);
+	void vram_w(offs_t offset, uint16_t data, uint16_t mem_mask = ~0);
+	uint16_t vram_r(offs_t offset);
+	void control_w(offs_t offset, uint16_t data, uint16_t mem_mask = ~0);
+	void paletteram_w(offs_t offset, uint16_t data, uint16_t mem_mask = ~0);
+	uint16_t dma_r(offs_t offset);
+	void dma_w(offs_t offset, uint16_t data, uint16_t mem_mask = ~0);
 	TMS340X0_TO_SHIFTREG_CB_MEMBER(to_shiftreg);
 	TMS340X0_FROM_SHIFTREG_CB_MEMBER(from_shiftreg);
 	TMS340X0_SCANLINE_IND16_CB_MEMBER(scanline_update);
-	DECLARE_MACHINE_RESET(midyunit);
-	DECLARE_VIDEO_START(midzunit);
 	DECLARE_VIDEO_START(midyunit_4bit);
 	DECLARE_VIDEO_START(midyunit_6bit);
-	DECLARE_VIDEO_START(common);
 	TIMER_CALLBACK_MEMBER(dma_callback);
 	TIMER_CALLBACK_MEMBER(autoerase_line);
 
+	void dma_draw(uint16_t command);
+	void init_gfxrom(int bpp);
+	void install_hidden_ram(mc6809e_device &cpu, int prot_start, int prot_end);
+
 	void main_map(address_map &map);
 
-	void dma_draw(uint16_t command);
+	void yunit_core(machine_config &config);
+	void yunit_4bpp(machine_config &config);
+	void yunit_6bpp(machine_config &config);
+};
+
+class midzunit_state : public midyunit_base_state
+{
+public:
+	midzunit_state(const machine_config &mconfig, device_type type, const char *tag)
+		: midyunit_base_state(mconfig, type, tag)
+		, m_narc_sound(*this, "narcsnd")
+	{
+	}
+
+	void zunit(machine_config &config);
+
+	void init_narc();
+
+	int narc_talkback_strobe_r();
+	DECLARE_CUSTOM_INPUT_MEMBER(narc_talkback_data_r);
+
+protected:
+	virtual void machine_reset() override;
+	virtual void video_start() override;
+
+private:
+	required_device<williams_narc_sound_device> m_narc_sound;
+
+	void narc_sound_w(offs_t offset, uint16_t data, uint16_t mem_mask = ~0);
+
+	void zunit_main_map(address_map &map);
+};
+
+class midyunit_cvsd_state : public midyunit_base_state
+{
+public:
+	midyunit_cvsd_state(const machine_config &mconfig, device_type type, const char *tag)
+		: midyunit_base_state(mconfig, type, tag)
+		, m_cvsd_sound(*this, "cvsd")
+	{
+	}
+
+	void yunit_cvsd_4bit_fast(machine_config &config);
+	void yunit_cvsd_4bit_slow(machine_config &config);
+	void yunit_cvsd_6bit_slow(machine_config &config);
+
+	void init_hiimpact();
+	void init_shimpact();
+	void init_smashtv();
+	void init_strkforc();
+	void init_trog();
+
+protected:
+	virtual void machine_reset() override;
+
+	optional_device<williams_cvsd_sound_device> m_cvsd_sound;
```

Comment:
You use this in `machine_reset` without checking that it’s present.  Should it be `required_device`?